### PR TITLE
diskann support new data type(fp16/bf16)

### PR DESCRIPTION
**Pull request by cqy123456**
issue: https://github.com/zilliztech/knowhere/issues/287
related milvus pr：https://github.com/milvus-io/milvus/pull/30716 , sift1m result also can see https://github.com/milvus-io/milvus/pull/30716
knowhere-test pr: https://github.com/milvus-io/knowhere-test/pull/176

### CODE REVIEW

File: src/index/diskann/diskann.cc [hunk 1]
```diff
@@ -32,7 +32,8 @@
 namespace knowhere {
 template <typename DataType>
 class DiskANNIndexNode : public IndexNode {
-    static_assert(std::is_same_v<DataType, fp32>, "DiskANN only support float");
+    static_assert(KnowhereFloatTypeCheck<DataType>::value,
+                  "DiskANN only support floating point data type(float32, float16, brain_float16)");
```
Comment (alexanderguzhva) [inline src/index/diskann/diskann.cc:36]: Please use `bfloat16` instead of `brain_float16` :)

File: tests/ut/test_diskann.cc [hunk 2]
```diff
@@ -355,29 +354,36 @@ TEST_CASE("Test DiskANN GetVectorByIds", "[diskann]") {
             json["data_path"] = kRawDataPath;
             json["max_degree"] = 5;
             json["search_list_size"] = kK;
-            json["pq_code_budget_gb"] = sizeof(float) * dim * kNumRows * 0.03125 / (1024 * 1024 * 1024);
+            json["pq_code_budget_gb"] = sizeof(DataType) * dim * kNumRows * 0.125 / (1024 * 1024 * 1024);
```
Comment (alexanderguzhva) [inline tests/ut/test_diskann.cc:357]: what is `0.125` magic number and why it got changed from `0.03125` (1/32th) ?

File: tests/ut/test_diskann.cc [hunk 3]
```diff
@@ -412,3 +418,15 @@ TEST_CASE("Test DiskANN GetVectorByIds", "[diskann]") {
     fs::remove_all(kDir);
     fs::remove(kDir);
 }
+
+TEST_CASE("Test DiskANN Base ", "[diskann]") {
```
Comment (alexanderguzhva) [inline tests/ut/test_diskann.cc:422]: maybe, a 3 tests for each of the data types instead?

File: tests/ut/test_diskann.cc [hunk 4]
```diff
@@ -412,3 +418,15 @@ TEST_CASE("Test DiskANN GetVectorByIds", "[diskann]") {
...
+    BaseSearchTest<knowhere::fp32>();
+    BaseSearchTest<knowhere::bf16>();
+    BaseSearchTest<knowhere::fp16>();
+}
+
+TEST_CASE("Test DiskANN GetVectorByIds", "[diskann]") {
```
Comment (alexanderguzhva) [inline tests/ut/test_diskann.cc:428]: same

File: thirdparty/DiskANN/include/diskann/utils.h [hunk 5]
```diff
@@ -583,6 +585,12 @@ namespace diskann {
   template<typename T>
   float prepare_base_for_inner_products(const std::string in_file,
                                         const std::string out_file) {
+    if (!knowhere::KnowhereFloatTypeCheck<T>::value) {
+      std::stringstream stream;
+      stream << "DiskANN currently only supports floating point data for IP."
```
Comment (alexanderguzhva) [inline thirdparty/DiskANN/include/diskann/utils.h:590]: I would interpret this message that only `float` is supported, but not `fp16` or `bf16`. Please mention `fp16` and `bf16` if they are also supported.

File: thirdparty/DiskANN/include/diskann/utils.h [hunk 6]
```diff
@@ -642,24 +650,30 @@ namespace diskann {
...
   template<typename T>
   std::vector<float> prepare_base_for_cosine(const std::string in_file,
-                                        const std::string out_file) {
-    LOG_KNOWHERE_DEBUG_
-        << "Pre-processing base file by normalizing";
+                                             const std::string out_file) {
+    if (!knowhere::KnowhereFloatTypeCheck<T>::value) {
+      std::stringstream stream;
+      stream
+          << "DiskANN currently only supports floating point data for Cosine."
```
Comment (alexanderguzhva) [inline thirdparty/DiskANN/include/diskann/utils.h:672]: same note for `float`, `fp16` and `bf16`

File: thirdparty/DiskANN/src/distance.cpp [hunk 7]
```diff
@@ -5,20 +5,30 @@ namespace diskann {
   template<typename T>
   DISTFUN<T> get_distance_function(diskann::Metric m) {
     if (m == diskann::Metric::L2) {
-      return [](const T* x, const T* y, size_t size) -> T {
+      return [](const T* x, const T* y, size_t size) -> float {
         float res = 0;
-        for (size_t i = 0; i < size; i++) {
-          res += ((float) x[i] - (float) y[i]) * ((float) x[i] - (float) y[i]);
+        auto  x_fp32 = std::make_unique<float[]>(size);
```
Comment (alexanderguzhva) [inline thirdparty/DiskANN/src/distance.cpp:10]: So, every invokation of distance calculation will trigger two vector allocations? O_o
Should I add fp16 and bf16 distance calculations into Faiss instead?

File: thirdparty/DiskANN/src/distance.cpp [hunk 8]
```diff
@@ -5,20 +5,30 @@ namespace diskann {
...
       };
-    } else if (m == diskann::Metric::INNER_PRODUCT) {
-      return [](const T* x, const T* y, size_t size) -> T {
+    } else if (m == diskann::Metric::INNER_PRODUCT ||
+               m == diskann::Metric::COSINE) {
+      return [](const T* x, const T* y, size_t size) -> float {
         float res = 0;
-        for (size_t i = 0; i < size; i++) {
-          res += (float) x[i] * (float) y[i];
+        auto  x_fp32 = std::make_unique<float[]>(size);
```
Comment (alexanderguzhva) [inline thirdparty/DiskANN/src/distance.cpp:23]: same

File: src/index/diskann/diskann.cc (hunk 1 above)
Comment (cqy123456) [reply src/index/diskann/diskann.cc]: updated

File: tests/ut/test_diskann.cc (hunk 4 above)
Comment (cqy123456) [reply tests/ut/test_diskann.cc]: updated

File: thirdparty/DiskANN/include/diskann/utils.h (hunk 5 above)
Comment (cqy123456) [reply thirdparty/DiskANN/include/diskann/utils.h]: updated

File: thirdparty/DiskANN/include/diskann/utils.h (hunk 6 above)
Comment (cqy123456) [reply thirdparty/DiskANN/include/diskann/utils.h]: updated

File: thirdparty/DiskANN/src/distance.cpp (hunk 7 above)
Comment (cqy123456) [reply thirdparty/DiskANN/src/distance.cpp]: fp16/bf16 distance function will implemented later, like faiss::fvec_inner_product and fvec_L2sqr.

File: thirdparty/DiskANN/include/diskann/utils.h [hunk 9]
```diff
@@ -583,6 +585,13 @@ namespace diskann {
   template<typename T>
   float prepare_base_for_inner_products(const std::string in_file,
                                         const std::string out_file) {
+    if (!knowhere::KnowhereFloatTypeCheck<T>::value) {
```
Comment (liliu-z) [inline thirdparty/DiskANN/include/diskann/utils.h:588]: do constexpr

**Pull request merged**